### PR TITLE
connect: Add TCP Fast Open support for Windows

### PR DESCRIPTION
**Pull request by bagder**
This is the former PR #3327, saved by Johannes Schindelin (@dscho), rebased, squashed
and pushed again.

Requires Windows 10 ver 1607 or newer

### CODE REVIEW

**Comment by bagder** [comment]
Thanks for the tumbs up. If you're also :+1: for actually merging this, I'm all ears. I personally cannot tell how the logic around the presence and use of this function is or should be!

File: lib/config-win32.h [hunk 1]
```diff
@@ -609,6 +609,12 @@ Vista
 #  endif
 #endif
 
+/* Define if you have the ConnectEx WinSock 2 extension */
+/* This requires a build target of WinXP or newer */
+#if defined(_WIN32_WINNT) && (_WIN32_WINNT >= 0x0501)
```
Comment (MarcelRaad) [inline lib/config-win32.h]: Why do we need this here at all if it's overridden in curl_setup.h anyway?

Comment (bagder) [reply lib/config-win32.h]: The curl_setup.h logic doesn't define `HAVE_CONNECTEX` though so if we  would remove this section the other one would need polish.

**Comment by dscho** [comment]
> If you're also 👍 for actually merging this, I'm all ears. I personally cannot tell how the logic around the presence and use of this function is or should be!

It is definitely not ready for merging in my mind. I [mentioned in the original PR how I think the runtime check should work](https://github.com/curl/curl/pull/3327#discussion_r241702597): simply by verifying that the `Curl_ConnextEx` function pointer is non-`NULL`. If the function cannot be found [when asked for](https://github.com/curl/curl/pull/3327/files#diff-993ed876aee37b2aaa454987600d2e60R149), it is safe to assume that it is not supported on the current system. Of course, it would also be good to make sure that this runtime check is not performed over and over again on system that do not support `ConnectEx()`, so the patch really would need a bit more work (and testing!).

**Comment by bagder** [comment]
I slapped the "needs update" label on this to mark that its not ready. I personally am not the right person to drive through the changed mentioned above so I'll leave this open for a while to see if someone else wants to take it through to the goal or if it doesn't happen in another several weeks I'll probably close and add a mention in the TODO.

**Pull request closed without merging**